src/packet_analysis/protocol/ip/conn_key/fivetuple/Factory.cc

```diff
@@ -12,44 +13,22 @@ namespace zeek::conn_key::fivetuple {
 zeek::ConnKeyPtr Factory::DoNewConnKey() const { return std::make_unique<zeek::IPConnKey>(); }
 
 zeek::expected<zeek::ConnKeyPtr, std::string> Factory::DoConnKeyFromVal(const zeek::Val& v) const {
-    static auto unexpected_conn_id = zeek::unexpected<std::string>{"invalid connection ID record encountered"};
     auto ck = NewConnKey();
     auto* ick = static_cast<zeek::IPBasedConnKey*>(ck.get());
     auto& pt = ick->PackedTuple();
-    const auto& vt = v.GetType();
 
-    if ( ! IsRecord(vt->Tag()) )
-        return unexpected_conn_id;
+    if ( v.GetType() != id::conn_id )
+        return zeek::unexpected<std::string>{
+            util::fmt("expected conn_id, got %s", obj_desc_short(v.GetType()).c_str())};
 
-    auto* vr = vt->AsRecordType();
     auto vl = v.AsRecordVal();
 
     // Indices into conn_id's record field value list:
     int orig_h = 0, orig_p = 1, resp_h = 2, resp_p = 3, proto = 4;
 
-    if ( vr != id::conn_id ) {
-        // While it's not a conn_id, it may have equivalent fields.
-        orig_h = vr->FieldOffset("orig_h");
-        resp_h = vr->FieldOffset("resp_h");
-        orig_p = vr->FieldOffset("orig_p");
-        resp_p = vr->FieldOffset("resp_p");
-        proto = vr->FieldOffset("proto");
-
-        // clang-format off
-        if ( orig_h < 0 || vr->GetFieldType(orig_h)->Tag() != TYPE_ADDR ||
-	     resp_h < 0 || vr->GetFieldType(resp_h)->Tag() != TYPE_ADDR ||
-	     orig_p < 0 || vr->GetFieldType(orig_p)->Tag() != TYPE_PORT ||
-	     resp_p < 0 || vr->GetFieldType(resp_p)->Tag() != TYPE_PORT ||
-	     proto < 0  || vr->GetFieldType(proto)->Tag() != TYPE_COUNT ) {
-            return unexpected_conn_id;
-        }
-        // clang-format on
-    }
-
     if ( ! vl->HasField(orig_h) || ! vl->HasField(resp_h) || ! vl->HasField(orig_p) || ! vl->HasField(resp_p) ||
-         ! vl->HasField(proto) ) {
-        return unexpected_conn_id;
-    }
+         ! vl->HasField(proto) )
```

Comment:
Given the type definitions, this should not be able to happen under any circumstances, right?

If that's the case - should this perhaps also assert / throw an fatal error? Seems like the only way this can happen is some kind of memory corruption.

Also - independently - should this also check for the ctx field?

Reply:
Yeah, good call. I think you're right that this can't be an issue if we enforce `conn_id`, but let me try it...